Update to TinkerPop 3.4.4

Tinkerpop 3.4.4 is released

@FlorianHockmann I think I am doing something wrong. I cannot properly test JanusGraph master branch with `mvn clean install -Dtest.skip.tp=false -Pdocker,solr7` even on the clean Ubuntu 19.4 with Maven 3.6.0, Docker 19.03.2, OpenJDK 1.8.0_222. The VM has 4 GB of RAM.
Are you able to test current master branch? Commit [8c1b8fcce135d745ae10e3285b2b13442f041567](https://github.com/JanusGraph/janusgraph/commit/8c1b8fcce135d745ae10e3285b2b13442f041567).

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.janusgraph.diskstorage.berkeleyje.BerkeleyFixedLengthKCVSTest
Loading time (ms): 281
Reading time (ms): 46
[WARNING] Tests run: 30, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 25.195 s - in org.janusgraph.diskstorage.berkeleyje.BerkeleyFixedLengthKCVSTest
[INFO] Running org.janusgraph.diskstorage.berkeleyje.BerkeleyKeyValueTest
[INFO] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 4.252 s - in org.janusgraph.diskstorage.berkeleyje.BerkeleyKeyValueTest
[INFO] Running org.janusgraph.diskstorage.berkeleyje.BerkeleyLogTest
[WARNING] Tests run: 14, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 69.578 s - in org.janusgraph.diskstorage.berkeleyje.BerkeleyLogTest
[INFO] Running org.janusgraph.diskstorage.berkeleyje.BerkeleyVariableLengthKCVSTest
Loading time (ms): 378
Reading time (ms): 87
[WARNING] Tests run: 30, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 26.982 s - in org.janusgraph.diskstorage.berkeleyje.BerkeleyVariableLengthKCVSTest
[INFO] Running org.janusgraph.graphdb.BerkeleyTransactionTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.591 s <<< FAILURE! - in org.janusgraph.graphdb.BerkeleyTransactionTest
[ERROR] longRunningTxShouldBeRolledBack{File}  Time elapsed: 14.55 s  <<< ERROR!
org.janusgraph.core.JanusGraphException: Could not rollback transaction due to exception
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)
Caused by: java.lang.IllegalStateException: Transaction 91 detected open cursors while aborting
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)

[INFO] Running org.janusgraph.graphdb.berkeleyje.BerkeleyGraphTest
Failed message add
Failed message add
Failed message add
Failed message add
Attempting to repair partially failed transaction [2@7f00010129889-osboxes2n::2019-10-08T14:14:00.957Z]
Attempting to repair partially failed transaction [5@7f00010129889-osboxes2n::2019-10-08T14:14:01.711Z]
Attempting to repair partially failed transaction [3@7f00010129889-osboxes2n::2019-10-08T14:14:01.680Z]
Attempting to repair partially failed transaction [4@7f00010129889-osboxes2n::2019-10-08T14:14:01.696Z]
[WARNING] Tests run: 114, Failures: 0, Errors: 0, Skipped: 11, Time elapsed: 446.762 s - in org.janusgraph.graphdb.berkeleyje.BerkeleyGraphTest
[INFO] Running org.janusgraph.graphdb.berkeleyje.BerkeleyOLAPTest
Execution time (ms) [200|20100]: 2289
Execution time (ms) [100|5050]: 1418
Execution time (ms) [100|5050]: 4627
Execution time (ms) [100|5050]: 1105
0
[INFO] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 102.185 s - in org.janusgraph.graphdb.berkeleyje.BerkeleyOLAPTest
[INFO] Running org.janusgraph.graphdb.berkeleyje.BerkeleyOperationCountingTest
Retrievals: 8200
Hits: 7602
Misses: 598
33.215	9.666	7.683
[WARNING] Tests run: 13, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 65.168 s - in org.janusgraph.graphdb.berkeleyje.BerkeleyOperationCountingTest
[INFO] Running org.janusgraph.graphdb.berkeleyje.BerkeleyPartitionGraphTest
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 71.116 s - in org.janusgraph.graphdb.berkeleyje.BerkeleyPartitionGraphTest
[INFO] Running org.janusgraph.graphdb.database.management.BerkeleyManagementTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.147 s - in org.janusgraph.graphdb.database.management.BerkeleyManagementTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   BerkeleyTransactionTest.longRunningTxShouldBeRolledBack:49 » JanusGraph Could ...
[INFO] 
[ERROR] Tests run: 233, Failures: 0, Errors: 1, Skipped: 17
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for JanusGraph: Distributed Graph Database 0.5.0-SNAPSHOT:
[INFO] 
[INFO] JanusGraph: Distributed Graph Database ............. SUCCESS [ 29.507 s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ....... SUCCESS [06:18 min]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ......... SUCCESS [49:09 min]
[INFO] JanusGraph CodePipelines CI: Distributed release testing. SUCCESS [04:02 min]
[INFO] janusgraph-server .................................. SUCCESS [ 30.250 s]
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database .. FAILURE [14:08 min]
[INFO] JanusGraph-CQL: Distributed Graph Database ......... SKIPPED
[INFO] JanusGraph-Cassandra: Distributed Graph Database ... SKIPPED
[INFO] JanusGraph-HBase: Parent Module .................... SKIPPED
[INFO] JanusGraph-HBase: Version-independent Core ......... SKIPPED
[INFO] JanusGraph-HBase: HBase Server Used by Maven Testing SKIPPED
[INFO] JanusGraph-HBase: 1.x and 2.x Compatibility Shim ... SKIPPED
[INFO] JanusGraph-HBase: Universal binary ................. SKIPPED
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support SKIPPED
[INFO] JanusGraph-Hadoop: Module .......................... SKIPPED
[INFO] JanusGraph-Bigtable: Google Cloud Bigtable driver .. SKIPPED
[INFO] JanusGraph-Lucene: Indexing Support ................ SKIPPED
[INFO] JanusGraph-Solr: Distributed Indexing Support ...... SKIPPED
[INFO] JanusGraph-All: Complete JanusGraph Distribution ... SKIPPED
[INFO] JanusGraph-Dist: Tar and Zip Archives .............. SKIPPED
[INFO] JanusGraph-Doc: Manual for JanusGraph .............. SKIPPED
[INFO] JanusGraph-Examples: Examples for JanusGraph ....... SKIPPED
[INFO] Example-Common: Common Graph Code for Examples ..... SKIPPED
[INFO] Example-BerkeleyJE: BerkeleyJE Storage, Lucene Index SKIPPED
[INFO] Example-Cassandra: Cassandra Thrift Storage, Elasticsearch Index SKIPPED
[INFO] Example-Cql: Cassandra CQL Storage, Elasticsearch Index SKIPPED
[INFO] Example-HBase: HBase Storage, Solr Index ........... SKIPPED
[INFO] Example-RemoteGraph: Example with RemoteGraph ...... SKIPPED
[INFO] Example-TinkerGraph: Example with TinkerGraph ...... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:14 h
[INFO] Finished at: 2019-10-08T10:21:53-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (default-test) on project janusgraph-berkeleyje: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/osboxes/github/janusgraph/janusgraph/janusgraph-berkeleyje/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :janusgraph-berkeleyje
```

**/github/janusgraph/janusgraph/janusgraph-berkeleyje/target/surefire-reports/org.janusgraph.graphdb.BerkeleyTransactionTest.txt** contains:
```
-------------------------------------------------------------------------------
Test set: org.janusgraph.graphdb.BerkeleyTransactionTest
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.591 s <<< FAILURE! - in org.janusgraph.graphdb.BerkeleyTransactionTest
longRunningTxShouldBeRolledBack{File}  Time elapsed: 14.55 s  <<< ERROR!
org.janusgraph.core.JanusGraphException: Could not rollback transaction due to exception
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)
Caused by: java.lang.IllegalStateException: Transaction 91 detected open cursors while aborting
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)
```

Did you have such problems? What did you do in this situation? Start the test again?

Not sure if it will help, but the failed test in my situation (https://github.com/JanusGraph/janusgraph/issues/1617#issuecomment-539577478) is caused by the added test in this PR #1676

@mad would you be able to check why that test is failing?
```
-------------------------------------------------------------------------------
Test set: org.janusgraph.graphdb.BerkeleyTransactionTest
-------------------------------------------------------------------------------
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 14.591 s <<< FAILURE! - in org.janusgraph.graphdb.BerkeleyTransactionTest
longRunningTxShouldBeRolledBack{File}  Time elapsed: 14.55 s  <<< ERROR!
org.janusgraph.core.JanusGraphException: Could not rollback transaction due to exception
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)
Caused by: java.lang.IllegalStateException: Transaction 91 detected open cursors while aborting
	at org.janusgraph.graphdb.BerkeleyTransactionTest.longRunningTxShouldBeRolledBack(BerkeleyTransactionTest.java:49)
```

With the fix provided by @mad Tinkerpop tests passes for `JanusGraph-BerkeleyJE: Distributed Graph Database`. Now I am facing errors in `JanusGraph-CQL: Distributed Graph Database`.

```
[ERROR]   StarGraphTest.shouldHandleSelfLoops Multiple Failures (2 failures)
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
[ERROR]   StarGraphTest.shouldHashAndEqualsCorrectly Multiple Failures (2 failures)
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
[ERROR]   StarGraphTest.shouldSerializeCorrectlyUsingGryo Multiple Failures (2 failures)
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
[ERROR]   StarGraphTest.shouldValidateThatOriginalAndStarVerticesHaveTheSameTopology Multiple Failures (2 failures)
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
	java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
[INFO] 
[ERROR] Tests run: 1739, Failures: 1735, Errors: 0, Skipped: 4
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] JanusGraph: Distributed Graph Database ............. SUCCESS [ 34.574 s]
[INFO] JanusGraph-Core: Core Library for JanusGraph ....... SUCCESS [04:31 min]
[INFO] JanusGraph-Backend-testutils: Collection of Backend tests SUCCESS [01:22 min]
[INFO] JanusGraph CodePipelines CI: Distributed release testing. SUCCESS [03:44 min]
[INFO] janusgraph-server .................................. SUCCESS [  0.281 s]
[INFO] JanusGraph-InMemory: InMemory Backend for JanusGraph SUCCESS [22:48 min]
[INFO] JanusGraph-Test: Test Suite for JanusGraph ......... SUCCESS [  1.198 s]
[INFO] JanusGraph-BerkeleyJE: Distributed Graph Database .. SUCCESS [  02:02 h]
[INFO] JanusGraph-CQL: Distributed Graph Database ......... FAILURE [16:42 min]
[INFO] JanusGraph-Cassandra: Distributed Graph Database ... SKIPPED
[INFO] JanusGraph-HBase: Parent Module .................... SKIPPED
[INFO] JanusGraph-HBase: Version-independent Core ......... SKIPPED
[INFO] JanusGraph-HBase: HBase Server Used by Maven Testing SKIPPED
[INFO] JanusGraph-HBase: 1.x and 2.x Compatibility Shim ... SKIPPED
[INFO] JanusGraph-HBase: Universal binary ................. SKIPPED
[INFO] JanusGraph-ElasticSearch: Distributed Indexing Support SKIPPED
[INFO] JanusGraph-Hadoop: Module .......................... SKIPPED
[INFO] JanusGraph-Bigtable: Google Cloud Bigtable driver .. SKIPPED
[INFO] JanusGraph-Lucene: Indexing Support ................ SKIPPED
[INFO] JanusGraph-Solr: Distributed Indexing Support ...... SKIPPED
[INFO] JanusGraph-All: Complete JanusGraph Distribution ... SKIPPED
[INFO] JanusGraph-Dist: Tar and Zip Archives .............. SKIPPED
[INFO] JanusGraph-Doc: Manual for JanusGraph .............. SKIPPED
[INFO] JanusGraph-Examples: Examples for JanusGraph ....... SKIPPED
[INFO] Example-Common: Common Graph Code for Examples ..... SKIPPED
[INFO] Example-BerkeleyJE: BerkeleyJE Storage, Lucene Index SKIPPED
[INFO] Example-Cassandra: Cassandra Thrift Storage, Elasticsearch Index SKIPPED
[INFO] Example-Cql: Cassandra CQL Storage, Elasticsearch Index SKIPPED
[INFO] Example-HBase: HBase Storage, Solr Index ........... SKIPPED
[INFO] Example-RemoteGraph: Example with RemoteGraph ...... SKIPPED
[INFO] Example-TinkerGraph: Example with TinkerGraph ...... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:52 h
[INFO] Finished at: 2019-11-20T19:13:29+00:00
[INFO] Final Memory: 117M/2590M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (tinkerpop-test) on project janusgraph-cql: There are test failures.
[ERROR] 
[ERROR] Please refer to /usr/src/janusgraph/janusgraph-cql/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :janusgraph-cql
```

@farodin91 for some reason the container for CQL doesn't start on my system. Do you know what could be the problem of this?

```
[INFO]  T E S T S^M
[INFO] -------------------------------------------------------^M
        ^[[37m^[[1mℹ︎ Checking the system...^[[0m^[[0m^M
        ^[[32m✔ Docker version should be at least 1.6.0^[[0m^M
        ^[[32m✔ Docker environment should have more than 2GB free disk space^[[0m^M
        ^[[37m^[[1mℹ︎ Checking the system...^[[0m^[[0m^M
        ^[[32m✔ Docker version should be at least 1.6.0^[[0m^M
        ^[[32m✔ Docker environment should have more than 2GB free disk space^[[0m^M
[INFO] Running org.janusgraph.blueprints.cql.process.CQLComputerTest^M
[ERROR] Tests run: 718, Failures: 718, Errors: 0, Skipped: 0, Time elapsed: 192.083 s <<< FAILURE! - in org.janusgraph.blueprints.cql.process.CQLComputerTest^M
[ERROR] testMessagePassingOut  Time elapsed: 182.386 s  <<< FAILURE!^M
org.opentest4j.MultipleFailuresError: ^M
Multiple Failures (2 failures)^M
        org.testcontainers.containers.ContainerLaunchException: Container startup failed^M
        java.lang.NullPointerException: <no message>^M
^M
[ERROR] shouldSucceedWithProperTraverserRequirements  Time elapsed: 0.7 s  <<< FAILURE!^M
org.opentest4j.MultipleFailuresError: ^M
Multiple Failures (2 failures)^M
        java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager^M
        java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.cql.CachingCQLStoreManager
```

--- a/docs/basics/schema.md
+++ b/docs/basics/schema.md
@@ -221,6 +221,10 @@ implicitly created property keys have cardinality SINGLE and data type
 `Object.class`. Users can control automatic schema element creation by
 implementing and registering their own `DefaultSchemaMaker`.
 
+When defining a cardinality for a vertex property which differs from SINGLE, 
+the cardinality should be used for all values of the vertex property in the 
+first query (i.e. the query which defines a new vertex property key).
+
 It is strongly encouraged to explicitly define all schema elements and
 to disable automatic schema creation by setting `schema.default=none` in
 the JanusGraph graph configuration.

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -26,6 +26,45 @@ use the latest versions of the software.
 
 ## Release Notes
 
+### Version 0.4.1 (Release Date: January ?, 2020)
+
+```xml tab='Maven'
+<dependency>
+    <groupId>org.janusgraph</groupId>
+    <artifactId>janusgraph-core</artifactId>
+    <version>0.4.1</version>
+</dependency>
+```
+
+```groovy tab='Gradle'
+compile "org.janusgraph:janusgraph-core:0.4.1"
+```
+
+**Tested Compatibility:**
+
+-   Apache Cassandra 2.2.10, 3.0.14, 3.11.0
+-   Apache HBase 1.2.6, 1.3.1, 1.4.10, 2.1.5
+-   Google Bigtable 1.3.0, 1.4.0, 1.5.0, 1.6.0, 1.7.0, 1.8.0, 1.9.0, 1.10.0, 1.11.0
+-   Oracle BerkeleyJE 7.5.11
+-   Elasticsearch 5.6.14, 6.0.1, 6.6.0
+-   Apache Lucene 7.0.0
+-   Apache Solr 7.0.0
+-   Apache TinkerPop 3.4.4
+-   Java 1.8
+
+For more information on features and bug fixes in 0.4.1, see the GitHub milestone:
+
+-   <https://github.com/JanusGraph/janusgraph/milestone/15?closed=1>
+
+#### Upgrade Instructions
+
+##### TinkerPop: Upgrade from 3.4.1 to 3.4.4
+
+Adding multiple values in the same query to a new vertex property without explicitly defined type 
+(i.e. using `Automatic Schema Maker` to create a property type) requires explicit usage of `VertexProperty.Cardinality` 
+for each call (only for the first query which defines a property) if the `VertexProperty.Cardinality` is different than 
+`VertexProperty.Cardinality.single`.
+
 ### Version 0.4.0 (Release Date: July 1, 2019)
 Legacy documentation: <https://old-docs.janusgraph.org/0.4.0/index.html>
 

--- a/mkdocs.yml
+++ b/mkdocs.yml
@@ -122,6 +122,6 @@ nav:
 extra:
     latest_version: 0.4.0
     snapshot_version: 0.4.1-SNAPSHOT
-    tinkerpop_version: 3.4.1
+    tinkerpop_version: 3.4.4
     hadoop2_version: 2.7.7
     jamm_version: 0.3.0

--- a/pom.xml
+++ b/pom.xml
@@ -69,7 +69,7 @@
     </distributionManagement>
     <properties>
         <titan.compatible-versions>1.0.0,1.1.0-SNAPSHOT</titan.compatible-versions>
-        <tinkerpop.version>3.4.1</tinkerpop.version>
+        <tinkerpop.version>3.4.4</tinkerpop.version>
         <junit-platform.version>1.5.1</junit-platform.version>
         <junit.version>5.5.1</junit.version>
         <mrunit.version>1.1.0</mrunit.version>
@@ -126,7 +126,7 @@
         <test.excluded.groups>MEMORY_TESTS,PERFORMANCE_TESTS,BRITTLE_TESTS</test.excluded.groups>
         <dependency.locations.enabled>false</dependency.locations.enabled>
         <cassandra-driver.version>3.7.2</cassandra-driver.version>
-        <testcontainers.version>1.11.4</testcontainers.version>
+        <testcontainers.version>1.12.4</testcontainers.version>
     </properties>
     <modules>
         <module>janusgraph-codepipelines-ci</module>

